ocl_kernels.cl

```diff
@@ -0,0 +1,1323 @@
+                if (v22 != 0) {
+                    switch (++inc) {
+                        case 1: v = v22; break;
+                        case 2: if (mcFirstIsZero(cs, 2)) v = v22; break;
+                        case 3: if (mcFirstIsZero(cs, 3)) v = v22; break;
+                        default: if (mcFirstIsZero(cs, 4)) v = v22;
+                    }
+                    cs = mcStepSeed(cs, st);
+                }
+
+                if (v != 4 && !mcFirstIsZero(cs, 3)) {
+                    v = 0;
+                }
+            }
+            break;
+        case 4:
+            break;
+        default:
+            if (v00 == 0 || v20 == 0 || v02 == 0 || v22 == 0) {
+                int64_t cs = getChunkSeed(ss, i+x, j+z);
+                if (mcFirstIsZero(cs, 5)) {
+                    v = 0;
+                }
+            }
+    }
+
+    out[i + j*w] = v;
+}
+
+__kernel void removeBrim(int4 dims, __constant int* in, __global int* out) {
+    int w = dims.s2;
+    int h = dims.s3;
+    int pW = w + 2;
+    int pH = h + 2;
+
+    // Frame in seed range
+    in = in + pW * pH * get_global_id(2);
+    out = out + w * h * get_global_id(2);
+
+    int i = (int)get_global_id(0);
+    int j = (int)get_global_id(1);
+    out[i + j*w] = in[(i+1) + (j+1)*pW];
+}
+
+__kernel void mapZoom(__constant struct Layer* layer, int layer_id, int4 dims, __constant int* in, __global int* out) {
+    int seed_offset = layer_id + (int)get_global_id(2) * L_NUM;
+    int i = (int)get_global_id(0);
+    int j = (int)get_global_id(1);
+    int x = dims.s0;
+    int z = dims.s1;
+    int w = dims.s2;
+    int h = dims.s3;
+    int st = (int)layer[seed_offset].startSalt;
+    int ss = (int)layer[seed_offset].startSeed;
+
+    int pX = x >> 1;
+    int pZ = z >> 1;
+    int pW = ((x + w) >> 1) - pX + 1;
+    int pH = ((z + h) >> 1) - pZ + 1;
+
+    // Frame in seed range
+    in = in + pW * pH * get_global_id(2);
+    out = out + w * h * get_global_id(2);
+
+    int newW = pW << 1;
+    int newH = pH << 1;
+
+    int v00 = in[(j+0)*pW + i];
+    int v01 = in[(j+1)*pW + i];
+    int v10 = in[(j+0)*pW + i+1];
+    int v11 = in[(j+1)*pW + i+1];
+
+    int chunkX = (i + pX) << 1;
+    int chunkZ = (j + pZ) << 1;
+
+    int cs = ss;
+    cs += chunkX;
+    cs *= cs * 1284865837 + 4150755663;
+    cs += chunkZ;
+    cs *= cs * 1284865837 + 4150755663;
+    cs += chunkX;
+    cs *= cs * 1284865837 + 4150755663;
+    cs += chunkZ;
+
+    int zz = 2 * j - (z & 1);
+    int xx = 2 * i - (x & 1);
+
+    bool z0 = zz >= 0 && zz < h;
+    bool z1 = zz + 1 < h;
+    bool x0 = xx >= 0 && xx < w;
+    bool x1 = xx + 1 < w;
+
+    if (z0 && x0) {
+        out[xx + zz * w] = v00;
+    }
+    if (z1 && x0) {
+        out[xx + (zz + 1) * w] = (cs >> 24) & 1 ? v01 : v00;
+    }
+    cs *= cs * 1284865837 + 4150755663;
+    cs += st;
+    if (z0 && x1) {
+        out[xx + 1 + zz * w] = (cs >> 24) & 1 ? v10 : v00;
+    }
+    int v;
+    if      (v10 == v01 && v01 == v11) v = v10;
```

Comment:
I did a few experiments one day trying to remove this branches, which are from `select_mode_or_random`. This is the alternative that worked better to me, why is about 25% faster that the "if cascade" on my CPU. I hope that the difference is bigger on a GPU but can't try that myself:

https://github.com/Badel2/slime_seed_finder/blob/9334b161bd4b7b7b8d7251e48623d5803707921c/benches/select_mode_or_random.rs#L118

```suggestion
    int cv00 = (v00 == v10) + (v00 == v01) + (v00 == v11);
    int cv10 = (v10 == v01) + (v10 == v11);
    int cv01 = v01 == v11;

    if cv00 > cv10 && cv00 > cv01 {
        v = v00;
    } else if cv10 > cv00 {
        v = v10;
    } else if cv01 > cv00 {
        v = v01;
    } else {
        // v = random
    }
```

Reply:
This looks great! The assembly look significantly better, if only for the CPU. I did some rudimentary testing with CUDA C, and I was surprised that the improvement was only minor for a GPU. After some digging I found that the nvcc compiler manages to reduce the branching for this part of the device code quite well on its own (at least better than gcc).